### PR TITLE
Switched to factory function

### PR DESCRIPTION
**Pull request by tschersach**
New shiny and more robust design. Tell me what you think @lukehoban.

### CODE REVIEW

**Comment by tschersach** [comment]
This should work but it doesn't. I'm investigating.

File: index.js [hunk 1]
```diff
@@ -23,18 +23,13 @@ const token = config.require("token")
...
+        const wrapper = () => {
+            return () => require("@iopipe/iopipe")({token})(func)
         }
+
+        const serialized = await origSerializeFunction(wrapper, {...args, isFactoryFunction: true});        
         return {
-            ...serialized,
-            text: lines.join("\n"),
+            ...serialized
         };
```
Comment (lukehoban) [inline index.js]: This can probably just return serialized instead of copying it into a new object.

**Review by lukehoban — APPROVED** [review]
This looks great!  There’s one additional step we could take. If you are interested in doing it great, else I can do that in a follow up PR.

**Comment by tschersach** [comment]
Like I said, it doesn't work and I can't find a way to make it work. The final exported function is not correctly invoked and I'm wondering if it's not on IOPipe side. I did a lot of tests but I'm currently not sure how I can make it work.

**Comment by lukehoban** [comment]
Pushed a commit which makes things work here for me - looks like there was an extra layer of `() => {}` in the original version.  Also did some further simplification while I was there.  @Tirke Want to validate this yourself as well?

**Comment by tschersach** [comment]
Yeah I did try this way but it doesn't work with this example :

```typescript
const hello = new aws.serverless.Function('hello', {
    func: (event1, context) => {
        console.log(event1)
    },
    policies: ['arn:aws:iam::aws:policy/AWSLambdaFullAccess']
})


const api = new serverless.apigateway.API('api', {
    routes: [
        {method: 'GET', path: '/hello', handler: hello.lambda}
    ]
})
```

Function invocation will timeout.

Generated code :

```javascript
function __f1() {
  return (function() {
    with({  }) {

return (event1, context) => {
        return console.log(event1);
    };

    }
  }).apply(undefined, undefined).apply(this, arguments);
}

function __f0() {
  return (function() {
    with({ token: "<token>", func: __f1 }) {

return () => require("@iopipe/iopipe")({ token })(func);

    }
  }).apply(undefined, undefined).apply(this, arguments);
}

exports.handler = __f0();
```


Repo example with the bucket on put function is working fine, only difference I can see is the presence of the `awaiter` call

```javascript

__f2.arguments = null;
__f2.caller = null;

function __f2() {
  return (function() {
    with({  }) {

return function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};

    }
  }).apply(undefined, undefined).apply(this, arguments);
}

function __f1() {
  return (function() {
    with({ __awaiter: __f2 }) {

return (ev) => __awaiter(this, void 0, void 0, function* () { return console.log(ev); });

    }
  }).apply(undefined, undefined).apply(this, arguments);
}

function __f0() {
  return (function() {
    with({ token: "<token>", func: __f1 }) {

return () => require("@iopipe/iopipe")({ token })(func);

    }
  }).apply(undefined, undefined).apply(this, arguments);
}

exports.handler = __f0();
```

**Comment by lukehoban** [comment]
@Tirke Your code works for me - except that the lambda doesn't return anything so the HTTP request times out (and it appears IOPipe doesn't record anything about the function invocation in this case, which is a little surprising).

If I change your code to this - it works:

```
const hello = new aws.serverless.Function('hello', {
    func: (event1, context, cb) => {
        console.log(event1)
        cb(null, {
            statusCode: 200,
            body: "Hello world",
        });
    },
    policies: ['arn:aws:iam::aws:policy/AWSLambdaFullAccess']
})
```

So I think the implementation in this PR is sound.

**Comment by tschersach** [comment]
Yep, I'm so used to the new async + direct return of node 8 lambdas that I forgot about the callback ... 
Still weird about IOPipe indeed. 

All good for me.

**Comment by lukehoban** [comment]
Pushed one more update to support incoming factoryFunctions as well.

**Comment by lukehoban** [comment]
Thanks @Tirke!

**Pull request merged**